src/sorcha/readers/CombinedDataReader.py

```diff
@@ -0,0 +1,124 @@
+        self.ephem_primary = ephem_primary
+
+    def add_ephem_reader(self, new_reader):
+        """Add a new reader for ephemeris data.
+
+        Parameters
+        ----------
+        new_reader (ObjectDataReader): The reader for a specific input file.
+        """
+        pplogger = logging.getLogger(__name__)
+        if self.ephem_reader is not None:
+            pplogger.error("ERROR: Ephemeris reader already set.")
+            sys.exit("ERROR: Ephemeris reader already set.")
+        self.ephem_reader = new_reader
+
+    def add_aux_data_reader(self, new_reader):
+        """Add a new object reader that corresponds to an auxiliary input data type..
+
+        Parameters
+        ----------
+        new_reader (ObjectDataReader): The reader for a specific input file.
+        """
+        self.aux_data_readers.append(new_reader)
+
+    def read_block(self, block_size=None, verbose=False, **kwargs):
+        """Reads in a set number of rows from the input, performs
+        post-processing and validation, and returns a data frame.
+
+        Parameters:
+        -----------
+        block_size (int, optional): the number of rows to read in.
+            Use block_size=None to read in all available data.
+            [Default = None]
+
+        verbose (bool, optional): use verbose logging.
+
+        Returns:
+        -----------
+        res_df (Pandas dataframe): dataframe of the combined object data.
+
+        """
+        pplogger = logging.getLogger(__name__)
+        verboselog = pplogger.info if verbose else lambda *a, **k: None
+
+        if self.ephem_reader is None:
+            pplogger.error("ERROR: No ephemeris reader provided.")
+            sys.exit("ERROR: No ephemeris reader provided.")
+        if len(self.aux_data_readers) == 0:
+            pplogger.error("ERROR: No auxiliary readers provided.")
+            sys.exit("ERROR: No auxiliary readers provided.")
+
+        # Load object IDs from the primary table.
+        if self.ephem_primary:
+            verboselog(f"Loading object IDs from: {self.ephem_reader.get_reader_info()}")
+            ephem_df = self.ephem_reader.read_rows(self.block_start, block_size)
+            self.block_start += len(ephem_df)
+
+            obj_ids = ephem_df["ObjID"].unique().tolist()
+        else:
+            verboselog(f"Loading object IDs from: {self.aux_data_readers[0].get_reader_info()}")
+            primary_df = self.aux_data_readers[0].read_rows(self.block_start, block_size)
+            self.block_start += len(primary_df)
+
+            obj_ids = primary_df["ObjID"].unique().tolist()
+            ephem_df = None
+
+        # No more data to load.
+        if len(obj_ids) == 0:
+            return None
+
+        # Load in the data for this block.
+        if ephem_df is None:
+            ephem_df = self.ephem_reader.read_objects(obj_ids)
+        ephem_ids = set(ephem_df["ObjID"].unique().tolist())
+
+        for reader in self.aux_data_readers:
```

Comment:
This for loop isn't actually doing another full read of the first aux_reader, right? I think that it's using the inverse index to read in only the required rows, but I wanted to make sure that my understanding is correct.

Reply:
That is correct. Outside the for loop, the program is reading in a set range of rows from `aux_data_readers[0]`. During the first iteration of the for loop, the program will re-open the file and load the rows for the matching object IDs.

That being said, this is still reopening the file and reading it back in. Let me update it to only do that if it has not already loaded the `aux_data_readers[0]` data.